contracts/ronin/random-beacon/RoninRandomBeacon.sol

```diff
@@ -0,0 +1,407 @@
+import { IRoninTrustedOrganization } from "../../interfaces/IRoninTrustedOrganization.sol";
+import { ISlashRandomBeacon } from "../../interfaces/slash-indicator/ISlashRandomBeacon.sol";
+import { LibSLA, RandomRequest } from "../../libraries/LibSLA.sol";
+import { TConsensus } from "../../udvts/Types.sol";
+import { ContractType } from "../../utils/ContractType.sol";
+import { ErrLengthMismatch, ErrUnauthorizedCall } from "../../utils/CommonErrors.sol";
+
+contract RoninRandomBeacon is
+  Initializable,
+  VRF,
+  HasContracts,
+  GlobalConfigConsumer,
+  PCUSortValidatorsByBeacon,
+  IRandomBeacon
+{
+  using LibSLA for uint256[2];
+
+  uint256[50] private __gap;
+
+  /// @dev The threshold of cooldown period for key hash change and newly register candidates.
+  uint256 private constant COOLDOWN_PERIOD_THRESHOLD = 1;
+
+  /// @dev Period of the beacon validator selection is activated.
+  uint256 private _activatedAtPeriod;
+  /// @dev The threshold of unavailability to slash.
+  uint256 private _unavailabilitySlashThreshold;
+  /// @dev Mapping of consecutive unavailable count per validator.
+  mapping(address gvId => uint256 count) private _unavailableCount;
+  /// @dev Mapping of beacon per period.
+  mapping(uint256 period => Beacon beacon) private _beaconPerPeriod;
+  /// @dev The maximum pick threshold for validator type.
+  mapping(ValidatorType validatorType => uint256 threshold) private _validatorThreshold;
+
+  constructor() {
+    _disableInitializers();
+  }
+
+  function initialize(
+    address profile,
+    address staking,
+    address trustedOrg,
+    address validatorSet,
+    address slashIndicator,
+    uint256 slashThreshold,
+    uint256 initialSeed,
+    uint256 activatedAtPeriod,
+    ValidatorType[] calldata validatorTypes,
+    uint256[] calldata thresholds
+  ) external initializer {
+    _activatedAtPeriod = activatedAtPeriod;
+    _setUnavailabilitySlashThreshold(slashThreshold);
+    _requestRandomSeed(activatedAtPeriod, initialSeed);
+    _bulkSetValidatorThresholds(validatorTypes, thresholds);
+
+    _setContract(ContractType.PROFILE, profile);
+    _setContract(ContractType.STAKING, staking);
+    _setContract(ContractType.VALIDATOR, validatorSet);
+    _setContract(ContractType.SLASH_INDICATOR, slashIndicator);
+    _setContract(ContractType.RONIN_TRUSTED_ORGANIZATION, trustedOrg);
+  }
+
+  /**
+   * @inheritdoc IRandomBeacon
+   */
+  function bulkSetValidatorThresholds(
+    ValidatorType[] calldata validatorTypes,
+    uint256[] calldata thresholds
+  ) external onlyAdmin {
+    _bulkSetValidatorThresholds(validatorTypes, thresholds);
+  }
+
+  /**
+   * @inheritdoc IRandomBeacon
+   */
+  function setUnavailabilitySlashThreshold(uint256 threshold) external onlyAdmin {
+    _setUnavailabilitySlashThreshold(threshold);
+  }
+
+  /**
+   * @inheritdoc IRandomBeacon
+   */
+  function fulfillRandomSeed(RandomRequest calldata req, Proof calldata proof) external {
```

Comment:
This method is too long, split down to smaller internal methods

Reply:
- check period
- check caller is candidate
- check proof
- save proof...

Reply:
I don't think we should necessarily split many internal functions when it only be used in one place. That would increase code size in general

Reply:
Readability over optimization. Always make it clear, before putting any over-optimization.

Reply:
Readability over optimization. Always make it clear, before putting any over-optimization.

Reply:
splitted in #29